x/interchainquery/keeper/msg_server.go

```diff
@@ -28,105 +28,143 @@ func NewMsgServerImpl(keeper Keeper) types.MsgServer {
 
 var _ types.MsgServer = msgServer{}
 
-func (k msgServer) SubmitQueryResponse(goCtx context.Context, msg *types.MsgSubmitQueryResponse) (*types.MsgSubmitQueryResponseResponse, error) {
-	ctx := sdk.UnwrapSDKContext(goCtx)
-	q, found := k.GetQuery(ctx, msg.QueryId)
+// check if the query requires proving; if it does, verify it!
+func (k Keeper) VerifyKeyProof(ctx sdk.Context, msg *types.MsgSubmitQueryResponse, q types.Query) error {
+	pathParts := strings.Split(q.QueryType, "/")
 
-	if found {
+	// the query does NOT have an associated proof, so no need to verify it.
+	if pathParts[len(pathParts)-1] != "key" {
+		return nil
+	} else {
+		// the query is a "key" proof query -- verify the results are valid by checking the proof!
+		if msg.ProofOps == nil {
+			return fmt.Errorf("unable to validate proof. No proof submitted")
+		}
+		connection, _ := k.IBCKeeper.ConnectionKeeper.GetConnection(ctx, q.ConnectionId)
 
-		// ABORT PROCESSING QUERY RESPONSE IF WE EXCEEDED THE TTL
-		k.Logger(ctx).Info(fmt.Sprintf("[ICQ Resp] query %s with ttl: %d, resp time: %d.", msg.QueryId, q.Ttl, ctx.BlockHeader().Time.UnixNano()))
-		curT, err := cast.ToUint64E(ctx.BlockTime().UnixNano())
+		msgHeight, err := cast.ToUint64E(msg.Height)
 		if err != nil {
-			return nil, err
+			return err
+		}
+		height := clienttypes.NewHeight(clienttypes.ParseChainID(q.ChainId), msgHeight+1)
+		consensusState, found := k.IBCKeeper.ClientKeeper.GetClientConsensusState(ctx, connection.ClientId, height)
+		if !found {
+			return fmt.Errorf("unable to fetch consensus state")
 		}
 
-		if q.Ttl < curT {
-			errMsg := fmt.Sprintf("[ICQ Resp] aborting query callback due to ttl expiry! ttl is %d, time now %d for query of type %s with id %s, on chain %s", q.Ttl, ctx.BlockHeader().Time.UnixNano(), q.QueryType, q.ChainId, msg.QueryId)
-			k.DeleteQuery(ctx, msg.QueryId)
-			k.Logger(ctx).Error(errMsg)
-			return &types.MsgSubmitQueryResponseResponse{}, nil
+		clientState, found := k.IBCKeeper.ClientKeeper.GetClientState(ctx, connection.ClientId)
+		if !found {
+			return fmt.Errorf("unable to fetch client state")
```

Comment:
might not be a bad idea to log these errors and wrap them with some context like we normally do.
You could make a new error type, something like `ErrInvalidICQProof`
And then each of these is of the form
```
errMsg := fmt.Sprintf("unable to fetch consensus state for Query ID (%s)", q.Id)
k.Logger(ctx).Error(errMsg)
sdkerrors.Wrapf(types.ErrInvalidICQProof, errMsg)
```